offline/online mode

kp-cat opened the pull request:
### Describe the purpose of your pull request

- `set_offline()` / `set_online()` methods to indicate whether the SDK is allowed to make HTTP calls or not. In 'offline' mode the SDK works from the cache only.

### Requirement checklist (only if applicable)

- [x] I have covered the applied changes with automated tests.
- [x] I have executed the full automated test set against my changes.
- [x] I have validated my changes against all supported platform versions.
- [x] I have read and accepted the [contribution agreement](https://github.com/configcat/legal/blob/main/contribution-agreement.md).

kp-cat committed:
offline/online mode

kp-cat committed:
offline mode integration test

randycoulman commented on lib/config_cat.ex:
````diff
@@ -134,6 +134,13 @@ defmodule ConfigCat do
+    ```elixir
+    {ConfigCat, [sdk_key: "YOUR SDK KEY", read_timeout_milliseconds: 5000]}
+    ```
+
````
Looks like this documentation/description is copy-pasted from another option and needs to be updated.

randycoulman commented on lib/config_cat/cache_policy/auto.ex:
```diff
@@ -88,12 +118,16 @@ defmodule ConfigCat.CachePolicy.Auto do
-
+    else
       :ok
     end
```
```suggestion
    if state.offline do
      :ok
    else
      with original <- Helpers.cached_config(state),
           :ok <- Helpers.refresh_config(state) do
        if config_changed?(state, original) do
          safely_call_callback(state[:on_changed])
        end
        :ok
      end
    end
```

randycoulman commented on lib/config_cat/cache_policy/behaviour.ex:
```diff
@@ -7,5 +7,8 @@ defmodule ConfigCat.CachePolicy.Behaviour do
   @type refresh_result :: CachePolicy.refresh_result()
 
   @callback get(id()) :: {:ok, Config.t()} | {:error, :not_found}
+  @callback is_offline(id()) :: :bollean
```
```suggestion
  @callback is_offline(id()) :: :boolean
```

randycoulman commented on test/config_cat/cache_policy/auto_test.exs:
```diff
@@ -183,6 +183,33 @@ defmodule ConfigCat.CachePolicy.AutoTest do
   end
 
+  describe "offline" do
+    test "dose not fetch config when offline mode is set", %{config: config} do
```
```suggestion
    test "does not fetch config when offline mode is set", %{config: config} do
```

randycoulman commented on test/config_cat/cache_policy/lazy_test.exs:
```diff
@@ -86,4 +86,26 @@ defmodule ConfigCat.CachePolicy.LazyTest do
   end
+
+  describe "offline" do
+    test "dose not fetch config when offline mode is set", %{config: config} do
```
```suggestion
    test "does not fetch config when offline mode is set", %{config: config} do
```

randycoulman commented on test/config_cat/cache_policy/manual_test.exs:
```diff
@@ -53,4 +53,26 @@ defmodule ConfigCat.CachePolicy.ManualTest do
   end
+
+  describe "offline" do
+    test "dose not fetch config when offline mode is set", %{config: config} do
```
```suggestion
    test "does not fetch config when offline mode is set", %{config: config} do
```

randycoulman commented on test/integration_test.exs:
```diff
@@ -54,6 +54,25 @@ defmodule ConfigCat.IntegrationTest do
              "This text came from ConfigCat"
   end
 
+  test "dose not fetch config when offline mode is set" do
```
```suggestion
  test "does not fetch config when offline mode is set" do
```

randycoulman reviewed (approved):
I think this looks OK; some typos and documentation to fix.

It's unfortunate that the same code has to be duplicated in all three cache policy implementations. `CachePolicy.Helpers` exists to try to eliminate duplication between the policies, so you could move some of the code there, but probably not enough to be really worth it.

It's almost possible to check the `offline` flag in `Helpers.refresh_config`, but `Lazy` would end up updating the `last_update` time, which is not likely what we'd want, so I don't think you can do that.

kp-cat committed:
Apply suggestions from code review

Co-authored-by: Randy Coulman <randy@randycoulman.com>

kp-cat committed:
offline docs update + review suggestions

kp-cat commented:
Thanks @randycoulman, I've fixed docs + typos.

What bothers me here is that when the auto policy is in offline mode, polling still happens. It doesn't actually perform the specific fetch, but the `polled_refresh` function is still called from time to time. It would be nice to be able to stop this in offline mode, but I couldn't manage to put the offline value in the state here:

``` elixir
  @impl GenServer
  def handle_info(:polled_refresh, state) do
    pid = self()

    Task.start_link(fn ->
      refresh(state)
      schedule_next_refresh(state, pid)
    end)

    {:noreply, state}
  end
```

Also, if we could stop the polling, when it starts again because the auto policy is back online, it would be nice if it didn't immediately send a fetch request to the servers, but only after the polling interval has elapsed. We solved this in other SDKs by storing the `fetch_time` in the cache, and always writing down when the config was downloaded, so we could decide whether a fetch is needed or not. If the config in the cache is old enough, we download the updated one.

Actually, we have a ticket for this:
```
Currently, the TTL for LazyLoad and the poll interval for AutoPoll are stored only in memory.
This leads to the behavior that each SDK init triggers a config.json download.
We should track the download times in the cache too so when someone uses a custom cache
the SDK can spare those initial HTTP requests.
```

It would be nice to do this as well, but I don't feel I have the power to tackle cache manipulation. 😄

randycoulman commented:
> What bothers me here is that when the auto policy is in offline mode, polling still happens. It doesn't actually perform the specific fetch, but the `polled_refresh` function is still called from time to time. It would be nice to be able to stop this in offline mode

> Also, if we could stop the polling, when it starts again because the auto policy is back online, it would be nice if it didn't immediately send a fetch request to the servers, but only after the polling interval has elapsed.

I think this shouldn't be too hard to get working. I don't have time to code it up myself right now, but here's an outline of the solution:

- The `offline` flag should already be in the state, so no need to put it there.
- In `handle_continue(:initial_fetch, state)`, first check the `offline` flag. If `offline` is `true`, then don't call either `refresh(state)` or `schedule_next_refresh(state)`. That way, the initial refresh won't happen, and we also won't schedule any more refreshes.
- In `handle_info(:polled_refresh, state)`, first check the `offline` flag. If `offline` is `true`, then don't call `Task.start_link` or anything inside its anonymous function.
- In `handle_call(:set_online, _from, state)`, call `schedule_next_refresh(state)` before returning. The first refresh won't happen until after the polling interval expires.
- (optional) Check the `offline` flag in `handle_call(:force_refresh, _from, state)` and don't call `refresh(state)` if it is `false`. If you make this change, then there is no longer a need to check the flag in `refresh` itself, because all of its callers will have already checked it.

randycoulman commented:
> Currently, the TTL for LazyLoad and the poll interval for AutoPoll are stored only in memory.
> This leads to the behavior that each SDK init triggers a config.json download.
> We should track the download times in the cache too so when someone uses a custom cache
> the SDK can spare those initial HTTP requests.

This is doable, but not without a breaking change to the `ConfigCache` behaviour, I think. And obviously, this would have no effect on the default `InMemoryCache`, since the last download time would also be stored in memory.

The basic outline of a solution would be:
- Add a new `@callback` to the `ConfigCache` behaviour for fetching the last update time. To avoid this being a breaking change, define the new function as `@optional_callback` instead, but then every caller would have to first check if the function existed with `function_exported?`. See [the docs](https://hexdocs.pm/elixir/1.14.4/typespecs.html#optional-callbacks) for more details.
- In the `InMemoryCache` module, modify the shape of the state to be a map with a `config` key and a `last_updated` key and update the various handlers appropriately. In `handle_call({:set, value}, _from, _state)`, store both the config and the current time to the state and return that.
- Modify config-fetching code (maybe in `CachePolicy.Helpers.refresh_config`?) to first check the last update time. If it comes back as `nil`, a timestamp that's old enough, or the function isn't implemented (see above), we do a fetch like normal. Otherwise, we can just return.
- This might allow you to eliminate the `needs_fetch?` logic in `CachePolicy.Lazy`.

randycoulman committed:
♻️ Don't schedule auto-refreshes when offline

Adjusts the logic in `CachePolicy.Auto` so that it does no work at all when offline. When it comes back online, the next fetch is scheduled to occur after the polling interval expires.

randycoulman commented:
@kp-cat I rebased on latest master and then made the changes that keep `CachePolicy.Auto` from doing any work at all when offline.

If you're OK with my changes, feel free to merge when you're ready.

randycoulman committed:
🔊 Add logging for online/offline mode
